### PR TITLE
Added an example with png rounded modules using gd

### PR DESCRIPTION
**Pull request by livingroot**
Result: 

Similar to:
https://github.com/chillerlan/php-qrcode/issues/127

### CODE REVIEW

**Comment by codemasher** [comment]
Oh wow, this is incredible! Just let me have a quick look over the commit. 🚀

File: examples/imageRoundedModules.png
Comment (codemasher) [inline examples/imageRoundedModules.png:1]: Can you please remove the example image from the directory? I might include a render to the documentation at some point.

File: examples/pngWithRoundedShapes.php [hunk 1]
```diff
@@ -0,0 +1,210 @@
+<?php
+
```
Comment (codemasher) [inline examples/pngWithRoundedShapes.php:2]: Please add a file header with your credit and a short description perhaps:

```
/**
 * <description>
 * 
 * @see https://github.com/chillerlan/php-qrcode/issues/127
 *
 * @created      17.09.2023
 * @author       <you, e.g. GH username/profile>
 * @copyright    2023 <you>
 * @license      MIT
 */
```

File: examples/pngWithRoundedShapes.php [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                $this->image,
+                $x2 - $rectsize,
+                $y2 - $rectsize,
+                $x2,
+                $y2,
+                $this->moduleValues[$M_TYPE]
+            );
+        }
+
+        if ($neighbours & (1 << 1)) { // neighbour сверху
```
Comment (codemasher) [inline examples/pngWithRoundedShapes.php:82]: i guess "сверху" means "top"

File: examples/pngWithRoundedShapes.php [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                $this->image,
+                $x2 - $rectsize,
+                $y1,
+                $x2,
+                $y1 + $rectsize,
+                $this->moduleValues[$M_TYPE]
+            );
+        }
+
+        if ($neighbours & (1 << 5)) { // neighbour снизу
```
Comment (codemasher) [inline examples/pngWithRoundedShapes.php:103]: and "снизу" means "bottom" then

File: examples/pngWithRoundedShapes.php [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            );
+        }
+
+        // ---------------------
+        // inner rounding
+        // ---------------------
+
+        if (!$this->matrix->check($x, $y)) {
+
+            if ($neighbours & 1 && $neighbours & (1 << 7) && $neighbours & (1 << 1)) {
```
Comment (codemasher) [inline examples/pngWithRoundedShapes.php:130]: PHPCS will yell at this (and all other) "unbracketed operation"

The many braces but clean way would be:
```php
if (($neighbours & 1) && ($neighbours & (1 << 7)) && ($neighbours & (1 << 1))){}
```

File: examples/pngWithRoundedShapes.php [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    'scale' => 20
+]);
+
+$qrcode = new QRCode($options);
+
+$img = $qrcode->render("https://www.youtube.com/watch?v=dQw4w9WgXcQ");
+
+header('Content-type: image/png');
+
+echo $img;
```
Comment (codemasher) [inline examples/pngWithRoundedShapes.php:210]: Other than the mentioned it's mostly just code formatting, but don't worry about that, I'll just run it through the formatter later to align it with the other examples

**Review by codemasher — CHANGES_REQUESTED** [review]

**Comment by codemasher** [comment]
Also, i noticed that your branch is like 6 months behind `dev-main` - could you please update and check if it still works for you (spoiler: it probably won't and I'm currently trying to find out why).

**Comment by codemasher** [comment]
Update: it's because the `setPixel()` method has ben renamed to `module()` for consistency across the several output classes.

File: examples/pngWithRoundedShapes.php [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+require_once __DIR__.'/../vendor/autoload.php';
+
+// --------------------
+// Class definition
+// --------------------
+
+class QRCustomPNG extends QRGdImage {
+
+    protected function setPixel(int $x, int $y, int $M_TYPE): void {
```
Comment (codemasher) [inline examples/pngWithRoundedShapes.php:18]: The `setPixel()` method has ben renamed to `module()` in the current `dev-main`.

File: examples/pngWithRoundedShapes.php [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                    $this->moduleValues[$M_TYPE | QRMatrix::IS_DARK]
+                );
+            }
+        }
+
+        imagefilledellipse(
+            $this->image,
+            (int)(($x * $this->scale) + ($this->scale / 2)),
+            (int)(($y * $this->scale) + ($this->scale / 2)),
+            (int)(2 * $this->options->circleRadius * $this->scale),
```
Comment (codemasher) [inline examples/pngWithRoundedShapes.php:183]: `$this->options->circleRadius` creates some funny (probably unintended) effects when it's set to a value other than 0.5 - I think hardcoding 0.5 would be the better option here.

**Comment by livingroot** [comment]
Thank you for your review, i will fix this issues. But i can't see `dev-main` branch. Did you mean `main`?

**Comment by codemasher** [comment]
Yes, `main` it is - `dev-main` is the composer name - a habit of mine i guess haha.

**Comment by livingroot** [comment]
All done! However, I am not sure about PHPCS complaining about single if conditions like `if ($neighbours & (1 << 3)) {`. I think it would be even worse for readability if this condition had additional brackets. Should I add them too?

**Comment by livingroot** [comment]
Also, I have checked that everything is working fine

**Comment by codemasher** [comment]
> All done! However, I am not sure about PHPCS complaining about single if conditions like `if ($neighbours & (1 << 3)) {`. I think it would be even worse for readability if this condition had additional brackets. Should I add them too?

I agree that this case is a bit weird, but there's no way to my knowledge to exclude these cases via the xml config. I have added them throughout the library for consistency, alternatively you can just disable the ckech for that line with a `// phpcs:ignore`.


> Also, I have checked that everything is working fine

That's great! Thank you for your contribution! 🚀

**Review by codemasher — APPROVED** [review]

**Pull request merged**